upgrade/upgrade-scripts/rootfs-cleanup

```diff
@@ -0,0 +1,161 @@
+#
+def dpkgcmp(vers1: str, op: str, vers2: str) -> bool:
+    return subprocess.run(["dpkg", "--compare-versions", vers1, op,
+                           vers2]).returncode == 0
+
+
+#
+# Helper to obtain the "current-version" of a rootfs filesytem/snapshot.
+#
+def version(rootfs: str) -> str:
+    return str(
+        subprocess.check_output([
+            "zfs", "get", "-Hpo", "value", "com.delphix:current-version", rootfs
+        ]).decode("UTF-8").strip())
+
+
+#
+# Helper class used when sorting rootfs filesystems and snapshots. This allows
+# us to use the "sorted" builtin, to build a list of all rootfs filesystems and
+# snapshots on the system, sorted baseed on the "current-version" property.
+#
+class rootfscmp(object):
+
+    def __init__(self, rootfs: str, *args: List[Any]):
+        self.rootfs = rootfs
+
+    def __lt__(self, other: Type["rootfscmp"]) -> bool:
+        return dpkgcmp(version(self.rootfs), "lt", version(other.rootfs))
+
+    def __gt__(self, other: Type["rootfscmp"]) -> bool:
+        return dpkgcmp(version(self.rootfs), "gt", version(other.rootfs))
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, rootfscmp):
+            return NotImplemented
+        return dpkgcmp(version(self.rootfs), "eq", version(other.rootfs))
+
+    def __le__(self, other: Type["rootfscmp"]) -> bool:
+        return dpkgcmp(version(self.rootfs), "le", version(other.rootfs))
+
+    def __ge__(self, other: Type["rootfscmp"]) -> bool:
+        return dpkgcmp(version(self.rootfs), "ge", version(other.rootfs))
+
+    def __ne__(self, other: object) -> bool:
+        if not isinstance(other, rootfscmp):
+            return NotImplemented
+        return dpkgcmp(version(self.rootfs), "ne", version(other.rootfs))
+
+
+def main() -> None:
+    snapshots = subprocess.check_output([
+        "zfs", "list", "-r", "rpool/ROOT", "-d", "2", "-Ho", "name", "-t",
+        "snapshot"
+    ]).decode("UTF-8").splitlines()
+
+    for snap in snapshots:
+        name = snap.split("@")[1]
+
+        #
+        # We filter out any container-delphix snapshots, as those removed by
+        # the calls to "rootfs-container delete", on the relevant filesystems.
+        #
+        if re.search("^container-delphix.\w{7}$", name):
+            snapshots.remove(snap)
+            continue
+
+        #
+        # Out of an abundance of caution, verify the snapshot name matches
+        # our expectations.
+        #
+        if not re.search("^execute-upgrade.\w{7}$", name):
+            raise SystemExit("unexpected snapshot name: '{}'".format(name))
+
+    filesystems = subprocess.check_output([
+        "zfs", "list", "-r", "rpool/ROOT", "-d", "1", "-Ho", "name", "-t",
+        "filesystem"
+    ]).decode("UTF-8").splitlines()
+
+    #
+    # We remove the first entry in the list, as that'll refer to the "rpool/ROOT"
+    # dataset, which is not versioned and also not eligible for deletion.
+    #
+    filesystems = filesystems[1:]
```

Comment:
nit: instead of relying on a specific ordering you can also do something like:
```
filesystems.remove('rpool/ROOT')
```

```
>>> filesystems = subprocess.check_output([
...     "zfs", "list", "-r", "rpool/ROOT", "-d", "1", "-Ho", "name", "-t",
...     "filesystem"
... ]).decode("UTF-8").splitlines()
>>> filesystems
['rpool/ROOT', 'rpool/ROOT/delphix.SBr1UZg']
>>> filesystems.remove('rpool/ROOT')
>>> filesystems
['rpool/ROOT/delphix.SBr1UZg']
```

Reply:
ah, right.. good point, I'll do that.